sql/plan/exchange.go

```diff
@@ -0,0 +1,309 @@
+package plan
+
+import (
+	"fmt"
+	"io"
+	"sync"
+
+	errors "gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+// ErrNoPartitionable is returned when no Partitionable node is found
+// in the Exchange tree.
+var ErrNoPartitionable = errors.NewKind("no partitionable node found in exchange tree")
+
+// Exchange is a node that can parallelize the underlying tree iterating
+// partitions concurrently.
+type Exchange struct {
+	UnaryNode
+	Parallelism int
+}
+
+// NewExchange creates a new Exchange node.
+func NewExchange(
+	parallelism int,
+	child sql.Node,
+) *Exchange {
+	return &Exchange{
+		UnaryNode:   UnaryNode{Child: child},
+		Parallelism: parallelism,
+	}
+}
+
+// RowIter implements the sql.Node interface.
+func (e *Exchange) RowIter(ctx *sql.Context) (sql.RowIter, error) {
+	var partitionable sql.Partitionable
+	Inspect(e.Child, func(n sql.Node) bool {
+		if p, ok := n.(sql.Partitionable); ok {
+			partitionable = p
```

Comment:
should we check that there is only one partitionable?

Reply:
I guess the rule will already do that, right?